lib/User/Manager.php

```diff
@@ -135,84 +128,514 @@ private function checkAccess() {
-		// Do we have custom search attributes for this configuration?
-		$search = $this->access->getConnection()->ldapAttributesForUserSearch;
-		if(!empty($search)) {
-			$attributes = array_merge($attributes, $search);
+		if($this->ocConfig->getSystemValue('enable_avatars', true) === true && !$minimal) {
+			// attributes may come with big payload.
+
+			$attributes['jpegphoto'] = true;
+			$attributes['thumbnailphoto'] = true;
 		}
 
-		$homeRule = $this->access->getConnection()->homeFolderNamingRule;
-		if(strpos($homeRule, 'attr:') === 0) {
-			$attributes[] = substr($homeRule, strlen('attr:'));
+		// we don't need the empty attribute
+		unset($attributes['']);
+
+		return array_keys($attributes);
+	}
+
+	/**
+	 * Gets an UserEntry from the LDAP server from a distinguished name
+	 * @param $dn
+	 * @return UserEntry
+	 * @throws DoesNotExsitOnLDAPException when the dn supplied cannot be found on LDAP
+	 */
+	public function getUserEntryByDn($dn) {
+		$result = $this->access->executeRead(
+			$this->getConnection()->getConnectionResource(),
+			$dn,
+			$this->getAttributes(),
+			'objectClass=*',
+			20);
+		if($result === false || $result['count'] === 0) {
+			throw new DoesNotExsitOnLDAPException($dn);
 		}
+		// Try to convert entry into a UserEntry
+		return $this->getFromEntry($result);
+	}
+
+	/**
+	 * @brief returns a User object by the DN in an ldap entry
+	 * @param array $ldapEntry the ldap entry used to prefill the user properties
+	 * @return \OCA\User_LDAP\User\UserEntry
+	 * @throws \Exception when connection could not be established
+	 * @throws \InvalidArgumentException if entry does not contain a dn
+	 * @throws \OutOfBoundsException when username could not be determined
+	 */
+	public function getFromEntry($ldapEntry) {
+		$this->checkAccess();
+		$userEntry = new UserEntry($this->ocConfig, $this->logger, $this->getConnection(), $ldapEntry);
+		$dn = $userEntry->getDN();
 
-		if(!$minimal) {
-			// attributes that are not really important but may come with big
-			// payload.
-			$attributes = array_merge($attributes, array(
-				'jpegphoto',
-				'thumbnailphoto'
-			));
+		if(!$this->access->isDNPartOfBase($dn, $this->getConnection()->ldapBaseUsers)) {
+			throw new \OutOfBoundsException("DN <$dn> outside configured base domains: ".print_r($this->getConnection()->ldapBaseUsers, true));
 		}
 
-		return $attributes;
+		$uid = $this->resolveUID($userEntry);
+		$userEntry->setOwnCloudUID($uid);
+
+		// cache entries
+		$this->usersByDN[$dn] = $userEntry;
+		$this->usersByUid[$uid] = $userEntry;
+
+		return $userEntry;
 	}
 
 	/**
-	 * @brief returns a User object by it's ownCloud username
-	 * @param string $id the DN or username of the user
-	 * @return \OCA\User_LDAP\User\User|null
+	 * @param $uid
+	 * @return UserEntry
 	 */
-	protected function createInstancyByUserName($id) {
-		$dn = $this->access->username2dn($id);
-		if($dn !== false) {
-			return $this->createAndCache($dn, $id);
+	public function getCachedEntry($uid) {
```

Comment:
I remember @jvillafanez having objections about exposing functions like getCache, as it exposes internal implementation.

Reply:
+1 . This shouldn't be needed.

My worry is the next action to take if this method returns null because the user isn't cached, but it exists in the backend.
* If we consider the user doesn't exists in the backend, we'll likely have a bug.
* If we have to fetch the user from the backend, it's better if a `getEntry` method would take care of this matter: check the cache first and then check the backend if the user isn't in the cache.
* If we rely on other method to be executed first to cache the user, this is a very bad approach because the method might not be executed in other scenarios, causing weird behaviour.

I don't see an use case where this is needed.

Reply:
Ok, so we are discussing details, while the code bit rots. I say we fix the lowercaseing of keys in the ldap entry and then have a test party to release ASAP. Improvements / refactoring can be done in subsequent PRs.